src/test/java/org/jenkinsci/plugins/workflow/graphanalysis/ForkScannerTest.java

```diff
@@ -992,4 +995,43 @@ public void testParallelCorrectEndNodeForVisitor() throws Exception {
         testParallelFindsLast(jobPauseSecond, "wait2");
         testParallelFindsLast(jobPauseMiddle, "wait3");
     }
+
+    @Test
+    public void parallelInParallelVisitsOuterParallelStartTwice() throws Exception {
+        WorkflowJob p = r.createProject(WorkflowJob.class);
+        p.setDefinition(new CpsFlowDefinition(
+                "stage('MyStage') {\n" +
+                "  parallel(\n" +
+                "    outerA: {\n" +
+                "      semaphore('outerA')\n" +
+                "    },\n" +
+                "    outerB: {\n" +
+                "      echo('outerB')\n" +
+                "    },\n" +
+                "    outerC: {\n" +
+                "      parallel(\n" +
+                "        innerA: {\n" +
+                "          semaphore('innerA')\n" +
+                "        },\n" +
+                "        innerB: {\n" +
+                "          echo('innerB')\n" +
+                "        }\n" +
+                "      )\n" +
+                "    }\n" +
+                "  )\n" +
+                "}\n", true));
+        WorkflowRun b = p.scheduleBuild2(0).waitForStart();
+        SemaphoreStep.waitForStart("outerA/1", b);
+        SemaphoreStep.waitForStart("innerA/1", b);
+        ForkScanner scanner = new ForkScanner();
+        scanner.setup(b.getExecution().getCurrentHeads());
+        List<String> visited = new ArrayList<>();
+        for (FlowNode node : scanner) {
+            visited.add(node.getId());
+        }
+        assertThat("Each FlowNode should be visited only once " + visited, visited.size(), equalTo(new HashSet<>(visited).size()));
```

Comment:
Right now, the parallel step start node, as well as the stage body start node, the stage step start node, and the the flow start node (!!) get visited more than once!